Provisioning end-to-end tests and API changes

CIPop opened the pull request:
<!--
Thank you for helping us improve the Azure IoT C SDK!

Here's a little checklist of things that will help it make its way to the repository: Note that you don't have to check all the boxes, we can help you with that. 
This being said, the more you do, the quicker it'll go through our gated build! 
--> 

# Checklist
- [X] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down). - Work in progress.
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [X] This pull-request is submitted against the `master` branch. 

    Adding common test configuration.
    Adding Code of conduct.
    E2E tests for provisioning client.
    Adding logging for both prod and test code.
    Updating API
    Removing Dice
    Changing HSM API to sync.
    Fixing Android sample.

    Adding SecurityClientX509
    Adding ProvisioningTransportHandle delegating pipeline.
    Refactoring HTTP and AMQP code.
    Adding chain CertificateInstaller to support Group Enrollment.
    TPM, X509 samples

<!-- How you solved the issue and the other things you considered and maybe rejected -->

tameraw commented on common/src/device/provisioning/transport/ClientApiVersionHelper.cs:
```diff
@@ -7,8 +7,5 @@ internal class ClientApiVersionHelper
     {
         public const string ApiVersionName = "api-version";
         public const string ApiVersion = "2017-08-31-preview";
```
>2017-08-31-preview [](start = 42, length = 18)

Need to change to GA version.

CIPop replied:
👍  dded TODO - I don't think another version is currently available on the service-side. /cc: @simonporter

tameraw commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs:
```diff
@@ -0,0 +1,111 @@
+        /// </summary>
+        /// <param name="deviceId">Device Identifier.</param>
+        public DeviceAuthenticationWithTokenRefresh(string deviceId, 
+            int timeToLiveBufferSeconds = 60) 
```
>timeToLiveBufferSeconds [](start = 16, length = 23)

I haven't gotten to where this is used yet. But if this is the SAS Expiry Time, should be 60min to line up with the rest of the SDKs.

CIPop replied:
Two points about this:
1. This is work-in-progress in this PR and isn't hooked-up to actually perform Token refresh. I am currently working on another PR for that. (Safe to ignore in this PR.)
2. Good point about the times. There are two actions here:
- remove the default parameters - bad API practice as they can't be versioned in the future.
- rewrite using the existing timeouts

tameraw commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs:
```diff
@@ -0,0 +1,111 @@
+        /// <summary>
+        /// Gets a snapshot expiry state.
+        /// </summary>
+        public bool IsExpiring => (_expiryTime - DateTime.UtcNow).TotalSeconds <= _timeToLiveBufferSeconds;
```
>IsExpiring [](start = 20, length = 10)

And renewal for the other SDKs is at 45 min.

CIPop commented on common/src/device/provisioning/transport/ClientApiVersionHelper.cs:
```diff
@@ -7,8 +7,5 @@ internal class ClientApiVersionHelper
         public const string ApiVersionName = "api-version";
         public const string ApiVersion = "2017-08-31-preview";
-
-        public const string ClientVersionName = "client-version";
```
This wasn't common code. Only AMQP needs it so I've moved it as-is in `AmqpClientLink.cs`.

CIPop commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs:
```diff
@@ -0,0 +1,111 @@
+    public abstract class DeviceAuthenticationWithTokenRefresh : IAuthenticationMethod
+    {
+        private string _deviceId;
+        private readonly int _timeToLiveBufferSeconds;
```
Yes, we need to make the code uniform. There are quite a few styles used right now.
We've picked https://github.com/dotnet/corefx/tree/master/Documentation#coding-guidelines

CIPop commented on provisioning/device/samples/ProvisioningDeviceClientTpm/Program.cs:
```diff
@@ -0,0 +1,84 @@
+    public static class Program
+    {
+        private static string s_idScope;
+        private const string RegistrationId = "testtpmregistration1";
```
Correct.

CIPop commented on provisioning/device/samples/ProvisioningDeviceClientX509/Program.cs:
```diff
@@ -0,0 +1,105 @@
+        //    For production code, it is advised that you install the certificate in the CurrentUser (My) store.
+        // DeviceID: iothubx509device1 (must match the CN part of the certificate Subject)
+
+        private static string s_idScope;
```
This is the unique identifier of a Provisioning Service. This is available from the portal but always needs to be supplied as a command-line parameter (or can be changed here).
`s_` is from the coding guidelines and identifies a local static.

CIPop commented on provisioning/device/src/ProvisioningDeviceClient.cs:
```diff
@@ -13,45 +13,65 @@ namespace Microsoft.Azure.Devices.Provisioning.Client
     public class ProvisioningDeviceClient
     {
-        private readonly string _globalDeviceEndpoint;
+        private const string DefaultGlobalDeviceEndpoint = "global.azure-devices-provisioning.net";
```
See rule 12 from https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md. 
`const`ants will use PascalCase. (i.e. they couldn't have been prefixes by `this.`)
`s_member`, `_member` are only used when `this.` could have been used.

CIPop commented on provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs:
```diff
@@ -0,0 +1,204 @@
+        private static readonly TimeSpan DefaultOperationPoolingIntervalMilliseconds = TimeSpan.FromSeconds(2);
+        private static readonly TimeSpan TimeoutConstant = TimeSpan.FromMinutes(1);
+        private const string OperationStatusAssigning = "assigning";
+        private const string OperationStatusUnassigned = "unassigned";
```
see above

tameraw commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTpm.cs:
```diff
@@ -0,0 +1,105 @@
+            return Task.FromResult<string>(sasToken);
+        }
+
+        private string AzureUrlEncode(string stringIn)
```
>AzureUrlEncode [](start = 23, length = 14)

Any reason why building your own instead of using one of the ones from the .Net lib?

CIPop replied:
Good point. @simonporter - is there anything special in the `AzureUrlEncode` different from the URI RFCs?

tameraw commented on device/samples/DeviceClientTwinSample/Program.cs:
```diff
@@ -6,6 +6,7 @@ namespace Microsoft.Azure.Devices.Client.Samples
     using System;
     using System.Threading.Tasks;
     using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.E2ETests;
```
>using Microsoft.Azure.Devices.E2ETests; [](start = 4, length = 39)

We shouldn't require this in our Samples.

CIPop replied:
The original code was reading a value from the `Environment`. Since it's hard to keep track of all environment variables and ensure everything is correctly set, I've unified the test configuration in `Configuration.IoTHub and Configuration.Provisioning`.

I will remove the configuration completely, letting the user to fill-in the correct required properties. 
I suggest we switch to command-line parameters to avoid mistakes where people check-in their SharedAccessKeys, Tokens, Hub addresses, etc.

tameraw commented on e2e/Microsoft.Azure.Devices.E2ETests.NetStandard/ProvisioningTests.cs:
```diff
@@ -0,0 +1,329 @@
+
+                _log.WriteLine($"{result.Status}");
+                
+                Assert.AreEqual(ProvisioningRegistrationStatusType.Assigned, result.Status);
```
>Assigned [](start = 67, length = 8)

If it fails, I wouldn't expect "Assigned".

CIPop replied:
Fixed.

tameraw commented on e2e/Microsoft.Azure.Devices.E2ETests.NetStandard/ProvisioningTests.cs:
```diff
@@ -0,0 +1,329 @@
+
+                var cts = new CancellationTokenSource(PassingTimeoutMiliseconds);
+
+                //security.RegistrationID = "INVALID";
```
>//security.RegistrationID = "INVALID"; [](start = 16, length = 38)

Looks like the test need to inject invalid RegistrationID (or bad cert or something) to cause the Registration failure.

CIPop replied:
Fixed.

tameraw commented on e2e/Microsoft.Azure.Devices.E2ETests.NetStandard/ProvisioningTests.cs:
```diff
@@ -0,0 +1,329 @@
+                _log.WriteLine($"{result.Status}");
+                _log.WriteLine($"ProvisioningClient AssignedHub: {result.AssignedHub}; DeviceID: {result.DeviceId}");
+
+                Assert.AreEqual(ProvisioningRegistrationStatusType.Assigned, result.Status);
```
>Assigned [](start = 67, length = 8)

Shouldn't also this be Unassigned?

CIPop replied:
The test is `ProvisioningDeviceClient_ValidRegistrationId_Register_Ok`: the expectation here is that after RegisterAsync completed, the device is provisioned.

CIPop replied:
Note: could be that the code shifted and this was meant for another test. I've corrected the E2E tests - PTAL.

tameraw commented on e2e/Microsoft.Azure.Devices.E2ETests.NetStandard/ProvisioningTests.cs:
```diff
@@ -0,0 +1,329 @@
+                _log.WriteLine($"{result.Status}");
+                _log.WriteLine($"ProvisioningClient AssignedHub: {result.AssignedHub}; DeviceID: {result.DeviceId}");
+
+                Assert.AreEqual(ProvisioningRegistrationStatusType.Assigned, result.Status);
```
>AreEqual [](start = 23, length = 8)

OR - Not Equal?

CIPop replied:
Must be equal - see above.

CIPop commented:
@tameraw @ryan-z-johnson I've submitted a fix for the TPM flow and addressed all your comments.
I've also switched the TPM sample to use MQTT as it was hitting #219 over AMQP.

I've successfully ran tests that are currently enabled (AMQP over TPM and WS doesn't work yet.)

ryan-z-johnson commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTokenRefresh.cs:
```diff
@@ -0,0 +1,111 @@
+
+            await _lock.WaitAsync();
+
+            if (!IsExpiring)
```
Why are we doing this twice? Is it to make sure that once we take application lock that it is still good?

CIPop replied:
It's a double checked lock pattern (see https://en.wikipedia.org/wiki/Double-checked_locking).

ryan-z-johnson commented on device/Microsoft.Azure.Devices.Client/DeviceAuthenticationWithTpm.cs:
```diff
@@ -0,0 +1,105 @@
+        protected override Task<string> SafeCreateNewToken(string iotHubName)
+        {
+            const long WINDOWS_TICKS_PER_SEC = 10000000;
+            const long EPOCH_DIFFERNECE = 11644473600;
```
is this meant to be hardcoded or will it get updated?

CIPop replied:
Hardcoded - Since we already have a token builder in the device client this will change with my next PR.

ryan-z-johnson commented on provisioning/device/samples/ProvisioningDeviceClientTpm/Program.cs:
```diff
@@ -0,0 +1,84 @@
+
+                Console.WriteLine(
+                    "In your Azure Device Provisioning Service please go to 'Manage enrollments' and select " +
+                    "'Individual Enrollments'. Select 'Add' then fill in the following:");
```
Is this required or can you do the whole flow besides initial creation of the resource via code?

CIPop replied:
Before this sample runs, it's required that the device is known by the service. That can be achieved by either the actions described above or, programmatic via ProvisioningServiceClient. These actions are expected to be performed in the hardware manufacturing factory.

tameraw commented:
>         /// TODO: Error code.

What's there TODO?

---
Refers to: provisioning/device/src/DeviceRegistrationResult.cs:73 in 9a7fb4b. [](commit_id = 9a7fb4bad4863c0527e3584051b7297e08fa7ea2, deletion_comment = False)

tameraw commented on provisioning/device/src/ProductInfo.cs:
```diff
@@ -0,0 +1,36 @@
+
+            // TODO:
+            // DO NOT EDIT the following line; it is updated by the bump_version script (https://github.com/Azure/iot-sdks-internals/blob/master/release/csharp/inputs.js)
+            const string Version = "1.5.2"; // CommonAssemblyVersion
```
>const string Version = "1.5.2 [](start = 12, length = 29)

Why do even need this. Can't the code extract directly from the Assembly file?

CIPop replied:
I was following the patterns from IoT Hub. That's a very good point. I'll add this to the TODO comment which was in place so we can update the bump_version script.

tameraw commented on provisioning/transport/amqp/src/AmqpAuthStrategyX509.cs:
```diff
@@ -0,0 +1,33 @@
+
+        public override Task OpenConnectionAsync(AmqpClientConnection connection, TimeSpan timeout, bool useWebSocket)
+        {
+            // TODO: install certificate chain.
```
>// TODO: install certificate chain. [](start = 12, length = 35)

What's this for?

CIPop replied:
It's actually done by the `ProvisioningDeviceClient.Create` factory at the moment but the code will be moved into the SecurityClientX509.

CIPop commented:
>   /// TODO: Error code.
> What's there TODO?

Removed. I've started a discussion regarding errorCode documentation.
